client/app/widgets/QuickCreate/index.jsx

```diff
@@ -29,129 +25,170 @@ export type State = {
   if (a.toLowerCase() > b.toLowerCase()) return 1;
   if (a.toLowerCase() < b.toLowerCase()) return -1;
   return 0;
 };
 
-const sortAlpha = (checkboxes: Checkbox[]): Checkbox[] =>
+export const sortAlpha = (checkboxes: Checkbox[]): Checkbox[] =>
   // eslint-disable-next-line implicit-arrow-linebreak
   checkboxes.sort((a: Checkbox, b: Checkbox) => alpha(a.label, b.label));
 
-export class QuickCreate extends React.Component<Props, State> {
-  constructor(props: Props) {
-    super(props);
-    this.state = {
-      checkboxes: sortAlpha(props.checkboxes),
-      open: false,
-      accordionOpen: props.checkboxes.some((cb) => cb.checked),
-    };
-  }
-
-  getCheckboxes = () => {
-    const { checkboxes } = this.state;
-    const checkboxesProp = [];
-    checkboxes.forEach((checkbox: Checkbox) => {
-      const checkboxProp = {
-        id: checkbox.id,
-        label: checkbox.label,
-        value: checkbox.value,
-        checked: checkbox.checked,
-      };
-      checkboxesProp.push(checkboxProp);
-    });
-    return checkboxesProp;
-  };
+export const labelExists = (checkboxes: Checkbox[], compareLabel: string) => checkboxes.filter(
+  (checkbox: Checkbox) => checkbox.label.toLowerCase() === compareLabel.toLowerCase(),
+).length;
 
-  labelExists = (label: string) => {
-    const { checkboxes } = this.state;
-    return checkboxes.filter(
-      (checkbox: Checkbox) => checkbox.label.toLowerCase() === label.toLowerCase(),
-    ).length;
-  };
+export const addToCheckboxes = (
+  { name, id, slug }: NewCheckbox,
+  checkboxes: Checkbox[],
+) => {
+  const newCheckboxes = checkboxes.slice(0);
+  newCheckboxes.push({
+    id: slug,
+    label: name,
+    value: id,
+    checked: true,
+  });
+  return sortAlpha(newCheckboxes);
+};
 
-  addToCheckboxes = (data: { name: string, id: string, slug: string }) => {
-    const { checkboxes } = this.state;
-    const { name, id, slug } = data;
-    const newCheckboxes = checkboxes.slice(0);
-    newCheckboxes.push({
-      id: slug,
-      label: name,
-      value: id,
-      checked: true,
-    });
-    return sortAlpha(newCheckboxes);
-  };
+// reducer types
+type OnChangePayload = {
+  label: string,
+  checkboxes: Checkbox[],
+  body: any,
+};
+type CheckboxData = {
+  name: string,
+  id: string,
+  slug: string,
+  success?: boolean,
+};
+type OnSubmitPayload = CheckboxData;
+type OnChangeAction = { type: 'ON_CHANGE', payload: OnChangePayload };
+type OnSubmitAction = { type: 'ON_SUBMIT', payload: OnSubmitPayload };
+type Action = OnChangeAction | OnSubmitAction;
+// reducer constants
+export const ON_CHANGE = 'ON_CHANGE';
+export const ON_SUBMIT = 'ON_SUBMIT';
 
-  onSubmit = (response: any) => {
-    const { data } = response;
-    if (data && data.success) {
-      this.setState({
+const quickCreateReducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case ON_CHANGE: {
+      return {
+        ...state,
+        checkboxes: action.payload.checkboxes,
```

Comment:
Question: is this using dot notation because of restrictions from Flow? I've run into some similar issues before. Airbnb [recommends](https://github.com/airbnb/javascript#destructuring) destructuring early to make values more visible, but I know it can be a hassle when dealing with types.

Reply:
Yeah fighting flow. I'll see if I can simplify at all.